finetuner-workflow/finetuner/evaluator.py

```diff
@@ -101,101 +53,113 @@ def get_gpu_ram() -> str:
-    for mod in modules:
-        original[mod] = mod.reset_parameters
-        mod.reset_parameters = dummy
-
-    with no_init_weights():
-        result = loading_code()
-    for mod in modules:
-        mod.reset_parameters = original[mod]
-
-    return result
-
-print(get_gpu_ram())
+print(MemoryUsage.now())
 
 start = time.time()
 
-tokenizer = AutoTokenizer.from_pretrained(args.tokenizer,
-                                          eos_token=args.eot,
-                                          pad_token=args.pad,
-                                          cache_dir=args.cache,
-                                          padding_side='left')
-
-model = AutoModelForCausalLM.from_pretrained(args.model,
-                                             # Can be a HuggingFace ID or directory.
-                                             cache_dir=args.cache,
-                                             use_cache=True)
-
-if args.fp16:
-    model = no_init(lambda: model.half().to(device))
-else:
-    model = no_init(lambda: model.to(device))
 
+tokens_to_add = {}
+if args.eot:
+    tokens_to_add["eos_token"] = args.eot
+if args.pad:
+    tokens_to_add["pad_token"] = args.pad
+
+tokenizer = AutoTokenizer.from_pretrained(
+    args.tokenizer,
+    **tokens_to_add,
+    cache_dir=args.cache,
+    padding_side="left"
+)
+
+tokens_to_add.clear()
+if "eos_token" not in tokenizer.special_tokens_map:
+    tokens_to_add["eos_token"] = "<|endoftext|>"
+if "pad_token" not in tokenizer.special_tokens_map:
+    tokens_to_add["pad_token"] = "<|endoftext|>"
+if tokens_to_add:
+    tokenizer.add_special_tokens(tokens_to_add)
+
+model = AutoModelForCausalLM.from_pretrained(
+    args.model,
+    # Can be a HuggingFace ID or directory.
+    cache_dir=args.cache,
+    use_cache=True
+)
+
+with no_init():
+    if args.fp16:
+        model = model.half().to(device)
+    else:
+        model = model.to(device)
+
+model.resize_token_embeddings(len(tokenizer))
 model.eval()
 
 
 duration = time.time() - start
 print(f"Loaded model in {duration:.2f}s")
-print(get_gpu_ram())
+print(MemoryUsage.now())
 torch.cuda.memory.empty_cache()
-print(get_gpu_ram())
-
-
-def evaluate(prompt,
-             generate_tokens: int = 200,
-             num_samples: int = 1,
-             eval_model: PreTrainedModel = model,
-             eval_tokenizer: PreTrainedTokenizer = tokenizer,
-             top_k: int = args.top_k,
-             top_p: float = args.top_p,
-             temperature: float = args.temperature,
-             repetition_penalty: float = args.repetition_penalty) -> List[str]:
+print(MemoryUsage.now())
+
+
+def evaluate(
+    prompt,
+    generate_tokens: int = 200,
+    num_samples: int = 1,
+    eval_model: PreTrainedModel = model,
+    eval_tokenizer: PreTrainedTokenizer = tokenizer,
+    top_k: int = args.top_k,
+    top_p: float = args.top_p,
+    temperature: float = args.temperature,
+    repetition_penalty: float = args.repetition_penalty
+) -> List[str]:
```

Comment:
To prevent drift going forward, should we just have the fintuner import `evaluate` from `evaluator.py`? Or at least have the `evaluate` function somewhere common between the two?

We'd have to move all the global level stuff in this script into a `main` function so it doesn't get run when importing `evaluate`, but I think that's better then making sure the functions across both files always match up during changes.

Thoughts?

Reply:
This was something I was already considering during this round of changes and here are my thoughts on it:

- On one hand, we could factor out nearly everything from `evaluator.py` except for the CLI—the model loading, the tokenizer setup, the `evaluate` function, and even the prompt file parsing are all shared with the finetuner. It would make the collection of the finetuner and evaluator scripts cleaner as a pair to share even more between each other.
  - It's also good to keep in mind that the shared model loading code will probably be getting more complex with the upcoming `tensorizer` integration, so separating that part out from the finetuner could just be cleaner in its own right.
- On the other hand, this code also serves as an example in this repository of writing an LLM finetuner and evaluator. Having standalone examples of the relevant ML logic presented step-by-step in each single file *could* potentially be more helpful to people who are only interested in one or the other.

Having the parts that pertain to evaluation be imported to the finetuner straight out of the evaluator also makes sense conceptually IMO. Though it also re-raises the question about whether or not to also deduplicate the model loading code, which conceptually could belong to either.

I don't have a strong opinion either way, so I'd defer to @wbrown for what he thinks would be better.